### ISSUE TITLE
[API Proposal]: Add `string.ReplaceAt`

### ISSUE THREAD

**Issue by Neme12** [dotnet/runtime#91124]
### Background and motivation

Currently, the only way to replace a character at a given index in a `string` is:
```c#
s.Substring(0, position) + newChar + s.Substring(position + 1)
```
It would be nice if there was a method on `string` to do this. That would not only be more convenient, but also much more efficient than the `Substring` approach. This is a popular question, see https://stackoverflow.com/questions/9367119/replacing-a-char-at-a-given-index-in-string as an example of someone looking for a `ReplaceAt` method (and the suggested solutions are even worse than the `Substring` approach, like converting the string to an array and back).

### API Proposal

```csharp
namespace System.Collections.Generic;

public class MyFancyCollection<T> : IEnumerable<T>
{
    public void Fancy(T item);
}
```

### API Usage

```csharp
// Fancy the value
var c = new MyFancyCollection<int>();
c.Fancy(42);

// Getting the values out
foreach (var v in c)
    Console.WriteLine(v);
```

### Alternative Designs

```diff
 namespace System;

 public sealed partial class String
 {
+    public string ReplaceAt(int index, char newValue);
+    public string ReplaceAt(int index, int count, string newValue);
 }
```

**Comment by Neme12** [comment]
@jkotas Why did you remove the "api-suggestion" label?

**Comment by jkotas** [comment]
> Currently, the most straightforward way to replace a character at a given index 

`$"{s.AsSpan(0, position)}{newChar}{s.AsSpan(position+1)}"` would be probably the best way to do this in modern .NET.

**Comment by Neme12** [comment]
That's true, but the vast majority people would just use `Substring` and might not even know about `AsSpan`.

**Comment by stephentoub** [comment]
> but the vast majority people would just use Substring and might not even know about AsSpan

And why would they be any more likely to know about a new ReplaceAt?

You've been opening a ton of API proposals. What's driving all of them?

**Comment by Neme12** [comment]
> And why would they be any more likely to know about and use a new ReplaceAt?

That's what I looked for and what it looks like other people look for as well, if you look at the stack overflow question.

> You've been opening a ton of API proposals. What's driving all of them?

Stuff I kept in the back of my mind for a long time as wanting to open issues for, until I finally got to doing it 😄

**Comment by Clockwork-Muse** [comment]
As an aside, I personally feel like a "replace at" type of API is a solution in search of a problem that was incorrectly misunderstood from some other issue (most likely an X-Y problem).

Perhaps the most relevant reason is that code should _rarely_ be working in "raw" characters, because in most real-world code you're going to be dealing with non-ASCII output, which would have problems with this (yes, `char` in C# is 16-bit, but most of the problems are going to come from outside the ASCII plane).

In most other common cases, the normal `Replace(...)` is likely to work fine.  Outside of that, you probably need to go most of the way down the stack to a raw byte array and twiddle down there.

**Comment by Neme12** [comment]
> Perhaps the most relevant reason is that code should rarely be working in "raw" characters

You're right, when I needed this, I needed to replace multiple characters, not just one, but I added the single-character version as well.

**Comment by Neme12** [comment]
The point is that sometimes you need to replace something at a given position and you don't need to look for it.

**Comment by skashtanoff** [comment]
Yes, sometimes we need it, but the `string`'s are immutable by design. If you need any "mutation", just create an extension method using span's that returns NEW string, as it mentioned above.
If you need to mutate the symbols too often, maybe you should choose an other container for `char`'s.

**Comment by tannergooding** [comment]
Closing based on the feedback above. StringBuilder is another viable alternative to this
